File: altgraph_data_app_java/src/main/java/org/cjoakim/cosmos/altgraph/data/common/graph/v2/JGraph.java
package org.cjoakim.cosmos.altgraph.data.common.graph.v2;

import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.cjoakim.cosmos.altgraph.data.common.DataConstants;
import org.cjoakim.cosmos.altgraph.data.common.ImdbConstants;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.struct.EdgeStruct;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.struct.EdgesStruct;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.struct.VertexValueStruct;
import org.jgrapht.GraphPath;
import org.jgrapht.alg.scoring.KatzCentrality;
import org.jgrapht.alg.scoring.PageRank;
import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
import org.jgrapht.graph.DefaultEdge;

import java.util.*;

/**
 * This class wrappers a JGraphT org.jgrapht.Graph object, and implements
 * graph traversal and lookup methods.
 * <p>
 * Chris Joakim, Microsoft, November 2022
 */

@Data
@Slf4j
public class JGraph implements DataConstants, ImdbConstants {

    private String domain;
    private String source;

    private Date refreshDate;
    private long refreshMs;

    private org.jgrapht.Graph<String, DefaultEdge> graph;

    public JGraph(String domain, String source) {
        super();
        this.domain = domain;
        this.source = source;
        refresh();
    }

    public int[] getVertexAndEdgeCounts() {
        int[] counts = new int[2];
        counts[0] = graph.vertexSet().size();
        counts[1] = graph.edgeSet().size();
        return counts;
    }

    /**
     * Find the shortest path with the DijkstraShortestPath class in JGraphT.
     */
    public GraphPath<String, DefaultEdge> getShortestPath(String v1, String v2) {
        log.warn("getShortestPath, v1: " + v1 + " to v2: " + v2);
        long start = System.currentTimeMillis();
        if (!isVertexPresent(v1)) {
            return null;
        }
        if (!isVertexPresent(v2)) {
            return null;
        }
        GraphPath<String, DefaultEdge> path =
                DijkstraShortestPath.findPathBetween(graph, v1, v2);
        long elapsed = System.currentTimeMillis() - start;

        if (path == null) {
            log.warn("path is null");
        } else {
            log.warn("elapsed milliseconds: " + elapsed);
            log.warn("path getLength:       " + path.getLength());
            log.warn("path getStartVertex:  " + path.getStartVertex());
            log.warn("path getEndVertex:    " + path.getEndVertex());
        }
        return path;
    }


    public EdgesStruct getShortestPathAsEdgesStruct(String v1, String v2) {

        long startMs = System.currentTimeMillis();
        GraphPath<String, DefaultEdge> path = getShortestPath(v1, v2);

        if (path != null) {
            EdgesStruct edgesStruct = new EdgesStruct();
            edgesStruct.setElapsedMs(System.currentTimeMillis() - startMs);
            edgesStruct.setVertex1(v1);
            edgesStruct.setVertex2(v2);
            for (DefaultEdge e : path.getEdgeList()) {
                EdgeStruct edgeStruct = parseDefaultEdge(e);
                edgesStruct.addEdge(edgeStruct);
            }
            return edgesStruct;
        } else {
            return null;
        }
    }

    public Set<DefaultEdge> edgesOf(String v) {
        if (isVertexPresent(v)) {
            return graph.edgesOf(v);
        }
        return null;
    }

    public JStarNetwork starNetworkFor(String rootVertex, int degrees) {
        JStarNetwork star = null;
        if (isVertexPresent(rootVertex)) {
            star = new JStarNetwork(rootVertex, degrees);

            for (int d = 1; d <= degrees; d++) {
                ArrayList<String> unvisitedList = star.getUnvisitedList();
                star.resetUnvisitedSet();

                log.warn("networkFor, unvisitedList size: " + unvisitedList.size() + " degree: " + d);
                for (int i = 0; i < unvisitedList.size(); i++) {
                    String v = unvisitedList.get(i);
                    Set<DefaultEdge> edges = graph.edgesOf(v);
                    star.addOutEdgesFor(v, edges, d);
                }
            }
        }
        star.finish();
        return star;
    }

    public Set<DefaultEdge> incomingEdgesOf(String v) {
        if (isVertexPresent(v)) {
            return graph.incomingEdgesOf(v);
        }
        return null;
    }

    public int degreeOf(String v) {
        if (isVertexPresent(v)) {
            return graph.degreeOf(v);
        }
        return -1;
    }

    public int inDegreeOf(String v) {
        if (isVertexPresent(v)) {
            return graph.inDegreeOf(v);
        }
        return -1;
    }

    public Double pageRankForVertex(String v) {
        if (isVertexPresent(v)) {
            PageRank pr = new PageRank(graph);
            return pr.getVertexScore(v);
        }
        return -1.0;
    }

    public Map pageRankForAll() {
        PageRank pr = new PageRank(graph);
        return pr.getScores();
    }

    public ArrayList<JRank> sortedPageRanks(int maxCount) {
        ArrayList<JRank> ranks = new ArrayList<JRank>();
        VertexValueStruct vvStruct = new VertexValueStruct();
        Map<String, Double> scores = this.pageRankForAll();
        Iterator<String> prAllIt = scores.keySet().iterator();
        while (prAllIt.hasNext()) {
            String vertex = prAllIt.next();
            Double value = scores.get(vertex);
            vvStruct.addRank(vertex, value);
        }
        vvStruct.sort();
        for (int i = 0; i < maxCount; i++) {
            ranks.add(vvStruct.getRank(i));
        }
        return ranks;
    }

    public Double centralityOfVertex(String v) {
        if (isVertexPresent(v)) {
            KatzCentrality kc = new KatzCentrality(graph);
            return kc.getVertexScore(v);
        }
        return -1.0;
    }

    public Map centralityRankAll() {
        KatzCentrality kc = new KatzCentrality(graph);
        return kc.getScores();
    }

    public void refresh() {
        long t1 = System.currentTimeMillis();
        org.jgrapht.Graph<String, DefaultEdge> newGraph = null;
        JGraphBuilder builder = new JGraphBuilder(source);
        log.warn("JGraph refresh(), domain: " + domain);

        try {
            if (domain.equalsIgnoreCase(GRAPH_DOMAIN_IMDB)) {
                newGraph = builder.buildImdbGraph();
                if (newGraph != null) {
                    refreshMs = System.currentTimeMillis() - t1;
                    refreshDate = new Date();
                    log.warn("JGraph refresh() - replacing graph with newGraph, elapsed ms: " + refreshMs);
                    graph = newGraph;
                }
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public boolean isVertexPresent(String v) {
        if (v != null) {
            return graph.containsVertex(v);
        }
        return false;
    }

    private EdgeStruct parseDefaultEdge(DefaultEdge e) {

        if (e != null) {
            String[] tokens = e.toString().split(":");
            if (tokens.length == 2) {
                EdgeStruct struct = new EdgeStruct();
                struct.setV1Value(tokens[0].replace('(', ' ').trim().intern());
                struct.setV2Value(tokens[1].replace(')', ' ').trim().intern());
                return struct;
            }
        }
        return null;
    }
}

File: dotnet/src/altgraph_shared_app/Services/Graph/v2/JGraph.cs
using altgraph_shared_app.Options;
using altgraph_shared_app.Services.Graph.v2.Structs;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using QuikGraph;
using QuikGraph.Algorithms;
using QuikGraph.Algorithms.Ranking;

namespace altgraph_shared_app.Services.Graph.v2
{
  public class JGraph : IJGraph
  {
    public string Domain { get; set; } = string.Empty;
    public string Source { get; set; } = string.Empty;
    public DateTime RefreshDate { get; set; }
    public long RefreshMs { get; set; }
    public IMutableGraph<string, Edge<string>>? Graph { get; set; } = null;
    private readonly ILogger _logger;
    private IJGraphBuilder _graphBuilder;
    private readonly ImdbOptions _imdbOptions;

    public event EventHandler<JGraphStartedLoadingProgressEventArgs>? JGraphStartedLoadingProgress;
    public event EventHandler<JGraphLoadingProgressEventArgs>? JGraphLoadingProgress;
    public event EventHandler<JGraphFinishedLoadingProgressEventArgs>? JGraphFinishedLoadingProgress;

    public JGraph(ILogger<JGraph> logger, IJGraphBuilder graphBuilder, IOptions<ImdbOptions> imdbOptions)
    {
      _logger = logger;
      _graphBuilder = graphBuilder;
      _graphBuilder.JGraphBuilderStartedLoadingProgress += OnJGraphStartedLoadingProgress;
      _graphBuilder.JGraphBuilderLoadingProgress += OnJGraphLoadingProgress;
      _graphBuilder.JGraphBuilderFinishedLoadingProgress += OnJGraphFinishedLoadingProgress;
      _imdbOptions = imdbOptions.Value;
      Domain = _imdbOptions.GraphDomain;
      Source = _imdbOptions.GraphSource;
    }

    public int[] GetVertexAndEdgeCounts()
    {
      int[] counts = new int[2];

      if (Graph is IVertexSet<string> graph0)
      {
        counts[0] = graph0.Vertices.Count();
      }

      if (Graph is IEdgeSet<string, Edge<string>> graph1)
      {
        counts[1] = graph1.Edges.Count();
      }

      return counts;
    }

    public IEnumerable<Edge<string>>? GetShortestPath(string v1, string v2)
    {
      _logger.LogDebug($"GetShortestPath, v1: {v1} to v2: {v2}");
      long start = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
      if (!IsVertexPresent(v1))
      {
        return null;
      }
      if (!IsVertexPresent(v2))
      {
        return null;
      }
      Func<Edge<string>, double> edgeCost = edge => 1; // Constant cost
      IEnumerable<Edge<string>>? path = null;

      if (Graph != null)
      {
        TryFunc<string, IEnumerable<Edge<string>>> tryGetPaths;

        if (Graph.IsDirected)
        {
          tryGetPaths = (Graph as IBidirectionalGraph<string, Edge<string>>).ShortestPathsDijkstra(edgeCost, v1);
        }
        else
        {
          tryGetPaths = (Graph as IUndirectedGraph<string, Edge<string>>).ShortestPathsDijkstra(edgeCost, v1);
        }
        long elapsed = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - start;

        if (!tryGetPaths(v2, out path))
        {
          _logger.LogWarning("path is null");
        }
        else
        {
          _logger.LogDebug($"elapsed milliseconds: {elapsed}");
          _logger.LogDebug($"path Count:       {path.Count()}");
          _logger.LogDebug($"path StartVertex:  {path.First()}");
          _logger.LogDebug($"path EndVertex:    {path.Last()}");
        }
      }
      return path;
    }


    public EdgesStruct? GetShortestPathAsEdgesStruct(string v1, string v2)
    {
      long startMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
      IEnumerable<Edge<string>>? path = GetShortestPath(v1, v2);

      if (path != null)
      {
        EdgesStruct edgesStruct = new EdgesStruct();
        edgesStruct.ElapsedMs = (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - startMs);
        edgesStruct.Vertex1 = v1;
        edgesStruct.Vertex2 = v2;
        foreach (Edge<string> e in path)
        {
          EdgeStruct? edgeStruct = ParseDefaultEdge(e);
          if (edgeStruct != null)
          {
            edgesStruct.AddEdge(edgeStruct);
          }
        }
        return edgesStruct;
      }
      else
      {
        return null;
      }
    }

    public HashSet<Edge<string>>? EdgesOf(string v)
    {
      if (IsVertexPresent(v))
      {
        HashSet<Edge<string>> edges = new HashSet<Edge<string>>();
        if (Graph != null)
        {
          if (Graph.IsDirected)
          {
            foreach (Edge<string> edge in ((IBidirectionalGraph<string, Edge<string>>)Graph).InEdges(v))
            {
              edges.Add(edge);
            }
            foreach (Edge<string> edge in ((IBidirectionalGraph<string, Edge<string>>)Graph).OutEdges(v))
            {
              edges.Add(edge);
            }
          }
          else
          {
            foreach (Edge<string> edge in ((IUndirectedGraph<string, Edge<string>>)Graph).AdjacentEdges(v))
            {
              edges.Add(edge);
            }
          }
        }
        return edges;
      }
      return null;
    }

    public JStarNetwork? StarNetworkFor(string rootVertex, int degrees)
    {
      JStarNetwork? star = null;
      if (IsVertexPresent(rootVertex))
      {
        star = new JStarNetwork(rootVertex, degrees);

        for (int d = 1; d <= degrees; d++)
        {
          List<string> unvisitedList = star.GetUnvisitedList();
          star.ResetUnvisitedSet();

          _logger.LogDebug($"networkFor, unvisitedList size: {unvisitedList.Count} degree: {d}");
          for (int i = 0; i < unvisitedList.Count; i++)
          {
            string v = unvisitedList[i];
            HashSet<Edge<string>>? edges = EdgesOf(v);
            if (edges != null)
            {
              star.AddOutEdgesFor(v, edges, d);
            }
          }
        }
      }
      star?.Finish();
      return star;
    }

    public HashSet<Edge<string>>? IncomingEdgesOf(string v)
    {
      if (Graph is IBidirectionalGraph<string, Edge<string>> && IsVertexPresent(v))
      {
        return ((IBidirectionalGraph<string, Edge<string>>)Graph)?.InEdges(v).ToHashSet();
      }
      return null;
    }

    public int DegreeOf(string v)
    {
      if (Graph != null && IsVertexPresent(v))
      {
        if (Graph.IsDirected)
        {
          return ((IBidirectionalGraph<string, Edge<string>>)Graph).Degree(v);
        }
        else
        {
          return ((IUndirectedGraph<string, Edge<string>>)Graph).AdjacentEdges(v).Count();
        }
      }
      return -1;
    }

    public int InDegreeOf(string v)
    {
      if (Graph != null && Graph is IBidirectionalGraph<string, Edge<string>> && IsVertexPresent(v))
      {
        return ((IBidirectionalGraph<string, Edge<string>>)Graph).InDegree(v);
      }
      return -1;
    }

    public double PageRankForVertex(string v)
    {
      if (Graph != null && Graph is IBidirectionalGraph<string, Edge<string>> && IsVertexPresent(v))
      {
        PageRankAlgorithm<string, Edge<string>> pr = new PageRankAlgorithm<string, Edge<string>>((IBidirectionalGraph<string, Edge<string>>)Graph);
        pr.Compute();
        return pr.Ranks[v];
        //return pr.GetVertexScore(v);
      }
      return -1.0;
    }

    public IDictionary<string, double>? PageRankForAll()
    {
      if (Graph != null && Graph is IBidirectionalGraph<string, Edge<string>>)
      {
        PageRankAlgorithm<string, Edge<string>> pr = new PageRankAlgorithm<string, Edge<string>>((IBidirectionalGraph<string, Edge<string>>)Graph);
        pr.MaxIterations = 100;
        pr.Damping = 0.85;
        pr.Tolerance = 1.0e-4;
        pr.Compute();
        return pr.Ranks;
      }
      return null;
    }

    public List<JRank> SortedPageRanks(int maxCount)
    {
      List<JRank> ranks = new List<JRank>();
      VertexValueStruct vvStruct = new VertexValueStruct();
      IDictionary<string, double>? scores = PageRankForAll();
      if (scores != null)
      {
        foreach (string vertex in scores.Keys)
        {
          double value = scores[vertex];
          vvStruct.AddRank(vertex, value);
        }
        vvStruct.Sort();
        for (int i = 0; i < maxCount; i++)
        {
          ranks.Add(vvStruct.GetRank(i));
        }
      }
      return ranks;
    }

    public double CentralityOfVertex(string v)
    {
      throw new NotImplementedException();
      // if (IsVertexPresent(v))
      // {
      //   KatzCentrality kc = new KatzCentrality(Graph);
      //   return kc.getVertexScore(v);
      // }
      // return -1.0;
    }

    public Dictionary<string, double> CentralityRankAll()
    {
      throw new NotImplementedException();
      // KatzCentrality kc = new KatzCentrality(Graph);
      // return kc.getScores();
    }

    public async Task RefreshAsync(bool directed = false)
    {
      long t1 = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
      IMutableGraph<string, Edge<string>>? newGraph = null;
      _logger.LogDebug($"JGraph refresh(), domain: {Domain}");

      try
      {
        if (Domain.Equals(_imdbOptions.GraphDomain, StringComparison.OrdinalIgnoreCase))
        {
          _graphBuilder.Directed = directed;
          newGraph = await _graphBuilder.BuildImdbGraphAsync();
          if (newGraph != null)
          {
            RefreshMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - t1;
            RefreshDate = DateTime.Now;
            _logger.LogDebug($"JGraph refresh() - replacing Graph with newGraph, elapsed ms: {RefreshMs}");
            Graph = newGraph;
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, ex.Message);
      }
    }

    public bool IsVertexPresent(string v)
    {
      if (Graph != null && v != null)
      {
        if (Graph.IsDirected)
        {
          return ((IBidirectionalGraph<string, Edge<string>>)Graph).ContainsVertex(v);
        }
        else
        {
          return ((IUndirectedGraph<string, Edge<string>>)Graph).ContainsVertex(v);
        }
      }
      return false;
    }

    private EdgeStruct? ParseDefaultEdge(Edge<string>? e)
    {
      if (e != null)
      {
        string[] tokens = e.ToString().Split(" -> ");
        if (tokens.Length == 2)
        {
          EdgeStruct edgeStruct = new EdgeStruct();
          edgeStruct.V1Value = tokens[0].Replace('(', ' ').Trim();
          edgeStruct.V2Value = tokens[1].Replace(')', ' ').Trim();
          return edgeStruct;
        }
      }
      return null;
    }

    private void OnJGraphStartedLoadingProgress(object? sender, JGraphBuilderStartedLoadingProgressEventArgs e)
    {
      OnRaiseJGraphStartedLoadingProgressEvent(new JGraphStartedLoadingProgressEventArgs(e.MaxCount));
    }

    private void OnJGraphLoadingProgress(object? sender, JGraphBuilderLoadingProgressEventArgs e)
    {
      OnRaiseJGraphLoadingProgressEvent(new JGraphLoadingProgressEventArgs(e.Progress));
    }

    private void OnJGraphFinishedLoadingProgress(object? sender, JGraphBuilderFinishedLoadingProgressEventArgs e)
    {
      OnRaiseJGraphFinishedLoadingProgressEvent(new JGraphFinishedLoadingProgressEventArgs(e.Count));
    }

    protected virtual void OnRaiseJGraphStartedLoadingProgressEvent(JGraphStartedLoadingProgressEventArgs e)
    {
      EventHandler<JGraphStartedLoadingProgressEventArgs>? raiseEvent = JGraphStartedLoadingProgress;

      if (raiseEvent != null)
      {
        raiseEvent(this, e);
      }
    }

    protected virtual void OnRaiseJGraphLoadingProgressEvent(JGraphLoadingProgressEventArgs e)
    {
      EventHandler<JGraphLoadingProgressEventArgs>? raiseEvent = JGraphLoadingProgress;

      if (raiseEvent != null)
      {
        raiseEvent(this, e);
      }
    }

    protected virtual void OnRaiseJGraphFinishedLoadingProgressEvent(JGraphFinishedLoadingProgressEventArgs e)
    {
      EventHandler<JGraphFinishedLoadingProgressEventArgs>? raiseEvent = JGraphFinishedLoadingProgress;

      if (raiseEvent != null)
      {
        raiseEvent(this, e);
      }
    }
  }

  public class JGraphStartedLoadingProgressEventArgs
  {
    public long MaxCount { get; private set; }
    public JGraphStartedLoadingProgressEventArgs(long maxCount)
    {
      MaxCount = maxCount;
    }
  }

  public class JGraphLoadingProgressEventArgs
  {
    public long Progress { get; private set; }
    public JGraphLoadingProgressEventArgs(long progress)
    {
      Progress = progress;
    }
  }

  public class JGraphFinishedLoadingProgressEventArgs
  {
    public long Count { get; private set; }
    public JGraphFinishedLoadingProgressEventArgs(long count)
    {
      Count = count;
    }
  }
}